File: _mcp/dedao-vault-mcp/dedao_vault_mcp/topics.py
import random

from .index import VaultIndex
from .pages import one_liner

VALID_MODES = {"random", "by_tag", "recently_updated", "least_linked", "unused"}


def pick_topics(index: VaultIndex, mode: str, type=None, tag=None,
                count: int = 5, written_covers=frozenset()):
    if mode not in VALID_MODES:
        raise ValueError(f"unknown mode: {mode}")
    cand = [
        p for p in index.pages.values()
        if (not type or p.type == type)
        and (not tag or tag in p.tags)
        and p.name not in written_covers
    ]
    if mode == "recently_updated":
        cand.sort(key=lambda p: str(p.frontmatter.get("updated", "")), reverse=True)
    elif mode == "least_linked":
        cand.sort(key=lambda p: (len(index.backlinks.get(p.name, set())), p.name))
    elif mode == "random":
        random.shuffle(cand)
    else:  # by_tag, unused
        cand.sort(key=lambda p: p.name)

    def why(p):
        if mode == "least_linked":
            return f"backlinks={len(index.backlinks.get(p.name, set()))}"
        if mode == "recently_updated":
            return f"updated={p.frontmatter.get('updated', '')}"
        return mode

    return [
        {"name": p.name, "type": p.type,
         "one_liner": one_liner(p.sections), "why": why(p)}
        for p in cand[:count]
    ]

File: _mcp/dedao-vault-mcp/src/topics.ts
import { VaultIndex } from "./vaultIndex";
import { Page, oneLiner } from "./pages";

export const VALID_MODES = new Set([
  "random", "by_tag", "recently_updated", "least_linked", "unused",
]);

function byName(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

export function pickTopics(
  index: VaultIndex,
  mode: string,
  type?: string,
  tag?: string,
  count = 5,
  writtenCovers: Set<string> = new Set(),
) {
  if (!VALID_MODES.has(mode)) throw new Error(`unknown mode: ${mode}`);
  const cand: Page[] = [];
  for (const p of index.pages.values()) {
    if (type && p.type !== type) continue;
    if (tag && !p.tags.includes(tag)) continue;
    if (writtenCovers.has(p.name)) continue;
    cand.push(p);
  }
  if (mode === "recently_updated") {
    cand.sort((a, b) => {
      const av = String(a.frontmatter.updated ?? "");
      const bv = String(b.frontmatter.updated ?? "");
      return av < bv ? 1 : av > bv ? -1 : 0;
    });
  } else if (mode === "least_linked") {
    const bc = (p: Page) => index.backlinks.get(p.name)?.size ?? 0;
    cand.sort((a, b) => bc(a) - bc(b) || byName(a.name, b.name));
  } else if (mode === "random") {
    for (let i = cand.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [cand[i], cand[j]] = [cand[j], cand[i]];
    }
  } else {
    cand.sort((a, b) => byName(a.name, b.name));
  }
  const why = (p: Page) => {
    if (mode === "least_linked") return `backlinks=${index.backlinks.get(p.name)?.size ?? 0}`;
    if (mode === "recently_updated") return `updated=${p.frontmatter.updated ?? ""}`;
    return mode;
  };
  return cand.slice(0, count).map((p) => ({
    name: p.name, type: p.type, one_liner: oneLiner(p.sections), why: why(p),
  }));
}
